cpp/src/indexing/point_quadtree.cu

```diff
@@ -0,0 +1,803 @@
+
+      return std::make_tuple(num_top_quads, quad_keys, quad_point_count);
+    }();
+
+    // Allocate this here so there's a chance RMM reuses
+    // the temporary memory allocated in the closure above
+    rmm::device_vector<uint32_t> quad_child_count(x.size());
+
+    cudf::size_type quad_tree_size{};
+    cudf::size_type num_parent_nodes{};
+    std::vector<cudf::size_type> b_pos{};
+    std::vector<cudf::size_type> e_pos{};
+
+    //
+    // Compute "full" quads for the tree at each level. Starting from the
+    // quadrant at the bottom (at the finest level of detail), aggregates the
+    // number of points and children in each quadrant per level. The key for
+    // each quadrant is each unique key from the level underneath it shifted
+    // right two bits (parent_key = child_key >> 2), as illustrated here:
+    //
+    // ------------------------------------------------------------------|
+    // initial point keys (sorted z-order Morton codes): [1, 3, 4, 7, 8] |
+    //                                                    ▾  ▾  ▾  ▾  ▾  |
+    //              .-------------------------------------`  :  :  :  :  |
+    //              :          .-----------------------------`  :  :  :  |
+    //              :          :          .---------------------`  :  :  |
+    //              :          :          :          .-------------`  :  |
+    //              :          :          :          :          .-----`  |
+    // ----------|--+----------+----------+----------+----------+--------|
+    //     level |        quadtree nodes [key, num children, num points] |
+    // ----------|--+----------+----------+----------+----------+--------|
+    //           |  ▾          ▾          ▾          ▾          ▾        |
+    // (start) 3 | [1, 0, 1]  [3, 0, 1]  [4, 0, 1]  [7, 0, 1]  [8, 0, 1] |
+    //           |     ▾          :          ▾          :          :     |
+    //           |     `---------.▾          `---------.▾          ▾     |
+    //         2 |            [0, 2, 2]             [1, 2, 2]  [2, 1, 1] |
+    //           |                ▾                     :          :     |
+    //           |                `--------------------.▾          ▾     |
+    //         1 |                                  [0, 2, 4]  [1, 1, 1] |
+    //           |                                      ▾          :     |
+    //           |                                      `---------.▾     |
+    //   (end) 0 |                                             [0, 2, 5] |
+    //                                                           (root)  |
+    //
+    std::tie(num_parent_nodes, quad_tree_size, b_pos, e_pos) =
+        construct_quadrants(num_levels, num_top_quads, quad_keys.begin(),
+                            quad_point_count.begin(), quad_child_count.begin(),
+                            stream);
+
+    // Shrink vectors' underlying device allocations to reduce peak memory usage
```

Comment:
Future optimization: don't bother shrinking if the data size is under a threshold (perhaps an option).